### PR TITLE
Large-memory routing: offload text >100KB to S3 (#497)

### PR DESCRIPTION
**Pull request by warlordofmars**
## Summary

Implements transparent routing of large text values to S3 storage, keeping DynamoDB items under the 400 KB limit. Text values exceeding 100 KB are automatically promoted to `text-large` type, uploaded to S3, and replaced with a pointer (`s3_uri`) in the DynamoDB META item. Smaller values remain inline unchanged, preserving backward compatibility with legacy items.

## Changes

- **New `BlobStore` class** (`src/hive/blob_store.py`): S3 wrapper for large-memory storage with `put()`, `get()`, and `delete()` operations. Defines `INLINE_TEXT_THRESHOLD_BYTES` (100 KB) and `MAX_BLOB_SIZE_BYTES` (10 MB) constants.

- **Storage routing** (`src/hive/storage.py`): 
  - Added `_route_large_value()` method that intercepts `put_memory()` calls and transparently offloads oversized text to S3
  - Lazy-initialized `blob_store` property to avoid requiring `HIVE_BLOBS_BUCKET` for tests that don't exercise the large-memory path
  - Preserves `size_bytes` on all text paths for future quota tracking (#500)

- **Memory model updates** (`src/hive/models.py`):
  - Added `value_type` (defaults to `"text"`), `s3_uri`, `content_type`, and `size_bytes` fields
  - Updated `to_dynamo_meta()` to only serialize non-default large-memory fields, ensuring legacy items round-trip byte-identical
  - Updated `from_dynamo()` to default `value_type` to `"text"` for pre-#497 items
  - Changed `value` field to allow `None` with empty-string default

- **Infrastructure** (`infra/stacks/hive_stack.py`):
  - Created `MemoryBlobsBucket` S3 bucket with SSE-S3 encryption, block-public-access, and tenant-scoped object keys
  - Granted MCP and API Lambda roles read/write/delete permissions on the blobs bucket
  - Added `HIVE_BLOBS_BUCKET` environment variable to Lambda functions

- **Server-side guards** (`src/hive/server.py`): Added `m.value or ""` fallbacks in `search_memories()` and `relate_memories()` to handle empty inline values until S3 fetch is wired in (#498)

## Test plan

- Added comprehensive unit tests in `tests/unit/test_blob_store.py` covering S3 operations, tenant isolation, and threshold constants
- Added integration tests in `tests/unit/test_storage.py` (`TestLargeMemoryRouting`) covering:
  - Small text staying inline with `size_bytes` captured
  - Large text (500 KB) routing to S3 with `text-large` type and empty inline value
  - Owner prefix fallback from `owner_user_id` to `owner_client_id` for pre-user-migration memories
  - Non-text types (image/blob) bypassing the router
  - Edge case of `None` value skipping upload
- Added model serialization tests in `tests/unit/test_models.py` verifying large-memory fields round-trip through DynamoDB and legacy items default correctly

All tests pass with 100% coverage maintained.

https://claude.ai/code/session_01Pws345BLe4hJdH2Qqrt7qb

### CODE REVIEW

File: src/hive/models.py [hunk 1]
```diff
@@ -62,6 +62,15 @@ class Memory(BaseModel):
...
     redacted_at: datetime | None = None  # when redact_memory tombstoned the value (#400)
+    # Large-memory foundation (#497). "text" remains the default and
+    # keeps the inline-value path unchanged. "text-large" offloads
+    # oversized text to S3 with ``value`` cleared and ``s3_uri`` set.
+    # "image" / "blob" are reserved for the binary remember_blob
+    # path that lands in #499.
+    value_type: str = "text"
+    s3_uri: str | None = None
+    content_type: str | None = None
+    size_bytes: int | None = None
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/models.py:73]: Fixed in 8a7e02c — changed `value_type` from `str` to `Literal["text", "text-large", "image", "blob"]` in `models.py`. Pydantic will now reject any out-of-range string at parse time.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/hive/server.py [hunk 2]
```diff
@@ -1231,7 +1231,10 @@ async def search_memories(
     now = datetime.now(timezone.utc)
     scored: list[tuple[Memory, float, float, float, float]] = []
     for m, sem in hydrated:
-        kw = keyword_score(query_tokens, m.value)
+        # value is None for the large-memory foundation (#497);
+        # #498 lands the S3 fetch. Until then, fall back to empty
+        # string so keyword_score stays typed as str.
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/server.py]: Fixed in 8a7e02c — rewrote the comment to accurately describe the empty-string placeholder: "Large-memory entries keep an empty-string inline placeholder until #498 lands the S3 fetch."

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/hive/server.py [hunk 3]
```diff
@@ -1231,7 +1231,10 @@ async def search_memories(
...
+        kw = keyword_score(query_tokens, m.value or "")
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/server.py:1237]: Fixed in 8a7e02c — raised `DEFAULT_MAX_VALUE_BYTES` from 10KB to 10MB, matching `MAX_BLOB_SIZE_BYTES`. Values in the 100KB–10MB range now reach the S3 routing path from MCP clients.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/hive/blob_store.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    """S3 wrapper for large-memory storage."""
+
+    def __init__(
+        self,
+        bucket_name: str | None = None,
+        region: str | None = None,
+        _s3_client: Any = None,
+    ) -> None:
+        # Read env at call time so tests can override after import.
+        self._bucket = bucket_name or os.environ.get("HIVE_BLOBS_BUCKET", "")
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/blob_store.py]: Fixed in 8a7e02c — `BlobStore.__init__` now raises `ValueError("BlobStore requires a non-empty bucket name; pass \`bucket_name\` or set the HIVE_BLOBS_BUCKET environment variable.")` when the resolved bucket name is empty.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/hive/storage.py [hunk 5]
```diff
@@ -200,6 +229,51 @@ def put_memory(self, memory: Memory, *, expected_version: str | None = None) ->
...
+        # Promote to text-large — write body to S3 under the
+        # workspace-equivalent prefix (user id today, workspace id
+        # post-#482).
+        owner = memory.owner_user_id or memory.owner_client_id
+        s3_uri = self.blob_store.put(
+            owner=owner,
+            memory_id=memory.memory_id,
+            body=encoded,
+            content_type="text/plain; charset=utf-8",
+        )
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/storage.py]: Fixed in 8a7e02c — `_route_large_value` now imports `MAX_BLOB_SIZE_BYTES` and raises `ValueError` when `len(encoded) > MAX_BLOB_SIZE_BYTES` before attempting the S3 upload.

---
_Generated by [Claude Code](https://claude.ai/code)_
